R/getPlot.R

```diff
@@ -0,0 +1,133 @@
+#'   smoothing. Defaults to 0 (no smoothing). Set to 1 for typical smoothing.
+#'   Higher values (up to 10) increase smoothing.
+#' - strokeThickness double The thickness of histogram and contour plot
+#'   lines. Defaults to 1.
+#' - tickLabelsQ logical Display tick labels. Defaults to false.
+#' - ticksQ logical Display ticks. Defaults to true.
+#' - width Integer Image width. Defaults to 228.
+#' - xAxisLabelQ logical Display x axis label. Overrides axisLabelsQ.
+#' - xAxisQ logical Display x axis line. Overrides axesQ.
+#' - xTickLabelsQ logical Display x tick labels. Overrides tickLabelsQ.
+#' - xTicksQ logical Display x ticks. Overrides ticksQ.
+#' - yAxisLabel logical Display y axis label. Overrides axisLabelsQ.
+#' - yAxisQ logical Display y axis line. Overrides axesQ.
+#' - yTickLabelsQ logical Display y tick labels. Overrides tickLabelsQ.
+#' - yTicksQ logical Display y ticks. Overrides ticksQ.
+#' @param destination character If specified, write the image to file, rather than
+#'   display it.
+#' @param overwrite logical If a destination is specified, allows destination
+#'   file to be overwritten.
+#' @param display logical Display the image. Defaults to `TRUE`.
+#'
+#' @md
+#' @export
+#' @examples
+#' \dontrun{
+#' parameters <- list(
+#'   "width" = 400,
+#'   "height" = 400,
+#'   "axesQ" = False,
+#'   "ticksQ" = False,
+#'   "color" = "#ff0000",
+#'   "renderGates" = True,
+#'   "preSubsampleN" = 4,
+#'   "preSubsampleP" = 0.5,
+#'   "postSubsampleN" = 4,
+#'   "postSubsampleP" = 0.5,
+#'   "seed" = 9,
+#'   "smoothing" = 0.75,
+#' )
+#' getPlot(experimentId,
+#'   fcsFileId,
+#'   plotType = "dot",
+#'   xChannel = "FSC-A",
+#'   yChannel = "FSC-W",
+#'   params = parameters
+#' )
+#' }
+getPlot <- function(experimentId,
+                    fcsFileId,
+                    plotType,
+                    xChannel,
+                    yChannel = "",
+                    zChannel = "",
+                    params = list(),
+                    destination = NULL,
+                    overwrite = FALSE,
+                    display = TRUE) {
+  checkDefined(experimentId)
+  experimentId <- lookupByName("experiments", experimentId)
+  checkDefined(fcsFileId)
+  fcsFileId <- lookupByName(paste("experiments", experimentId, "fcsfiles", sep = "/"), fcsFileId, "filename")
+
+  args <- list(
+    fcsFileId = fcsFileId,
+    plotType = plotType,
+    xChannel = xChannel,
+    yChannel = yChannel,
+    zChannel = zChannel,
+    axisLabelsQ = TRUE,
+    ticksQ = TRUE
+  )
+
+  args[names(params)] <- params
+  bytes <- baseGet(paste("experiments", experimentId, "plot", sep = "/"), args)
+
+  if (!display) {
+    return(invisible(bytes))
+  } else if (!is.null(destination)) {
+    if (!overwrite && file.exists(destination)) {
+      stop("Path exists and overwrite is FALSE")
+    }
+    if (!endsWith(destination, ".png")) destination <- paste0(path, ".png")
+    writeBin(bytes, destination)
+    return(invisible(bytes))
```

Comment:
This if/else chain can be tidied up like this:
```
if (!is.null(destination)) {
  ...
}

if (display) {
  if (is R studio) {
    ...
  } else if (is jupyter) {
    ...
  } else {
    warning("Not in RStudio or Jupyter. Can't display plot.")
  }
  return(invisible(bytes))
} else {
  return(bytes)
}
```

Reply:
:+1:

Reply:
This comment is still half pending. Lines 116, 124 and 127 should be deleted (the `return`s). Deleting 116 allows the plot to both be saved and displayed.

Reply:
Ah, okay, I put those returns in because I thought someone might want to display the plot and get the binary. Good point on deleting 116, that's useful.